### PR TITLE
Handle corrupted rocksdb database

### PR DESCRIPTION
**Pull request by lambda-conjecture**
This PR adds code to repair (or, in the case of ROCKSDB_LITE, delete) a corrupted db if one is encountered on startup.

On some testing we have done with OSX, we see regular occurrence of rocksdb database getting corrupted. The problem is, at that point, osquery itself is hosed - it won't restart until someone manually deletes the rocksdb database. 

While it's hard to generalize what's causing the corruption, in a few instances, we did catch this error:

I1214 10:05:25.787860 1064960 rocksdb.cpp:133] RocksDB: [ERROR] [events] [JOB 84] OpenCompactionOutputFiles for table #36365 fails at NewWritableFile with status IO error: /var/osquery/osquery.db/036365.sst: Permission denied
I1214 10:05:25.788249 1064960 rocksdb.cpp:133] RocksDB: [WARN] Compaction error: IO error: /var/osquery/osquery.db/036365.sst: Permission denied

It's not clear what suddenly causes the permission denied - osquery usually would have been running for multiple days before it runs into this. Nevertheless, it appears that we get into this state on average in about 10-12 days of continuously running osquery.

The test machines are people's actual work machines, and are being used as such - regular shutting of the lid, regular connecting/disconnecting to different networks (work, home, conference rooms etc.). They all have Sophos installed, as well as the usual battery of productivity apps.

We also see a persistent memory leak on osx (might be related to the queries we are running). On an average run, osquery worker will restart about every 2-3 days as it exceeds the watchdog memory threshold. This might be relevant, because rocksdb documentation FAQ specifically mentions that:
"Q: Is it safe to close RocksDB while another thread is issuing read, write or manual compaction requests?

A: No. The users of RocksDB need to make sure all functions have finished before they close RocksDB."
It may be that the kill request from the watcher hits at just the wrong time.

### CODE REVIEW

File: osquery/database/plugins/rocksdb.cpp [hunk 1]
```diff
@@ -243,6 +259,28 @@ void RocksDBDatabasePlugin::close() {
   }
 }
 
+void RocksDBDatabasePlugin::repairDB() {
+#if defined(ROCKSDB_LITE)
```
Comment (theopolis) [inline osquery/database/plugins/rocksdb.cpp]: We should assume `ROCKSDB_LITE` is always used.

File: osquery/database/plugins/rocksdb.cpp [hunk 2]
```diff
@@ -243,6 +259,28 @@ void RocksDBDatabasePlugin::close() {
...
+  if (s.ok()) {
+    for (auto file : file_names) {
+      fs::path file_path(file);
+      osquery::remove(file_path);
+    }
+  } else {
+    LOG(INFO) << "Unable to list " << path_ << ". Error " << s.toString();
+  }
+#else
+  auto rs = rocksdb::RepairDB(path_, options_, column_families_);
```
Comment (theopolis) [inline osquery/database/plugins/rocksdb.cpp]: Let's ping the RocksDB project and find out if this can be included in the LITE version.

**Comment by theopolis** [comment]
This is great, I'm also curious if the most-recent release still has this issue. Previously, before `2.2.0` when a table dropped privileges all of the process' threads would drop. If a RocksDB background thread created or accessed a new SST/LOG then it would be subject to the altered permissions. [Now](https://github.com/facebook/osquery/pull/2806) with `2.2.0` we only drop privileges for the thread executing the query.

File: osquery/database/plugins/rocksdb.cpp (hunk 1 above)
Comment (mshah-uptycs) [inline osquery/database/plugins/rocksdb.cpp]: I removed the #ifdef ROCKSDB_LITE construct

Comment (lambda-conjecture) [inline osquery/database/plugins/rocksdb.cpp]: Got it. Removed the #ifdef construct

**Comment by lambda-conjecture** [comment]
I should have an opportunity to try out 2.2.1 (or latest) at scale sometime in Mid-Jan. Having said that, I believe RocksDB is called in the context of the thread executing the query, yes? Unless rocksdb only creates new SST/LOG files in a background thread?

File: osquery/database/plugins/rocksdb.cpp [hunk 3]
```diff
@@ -243,6 +259,22 @@ void RocksDBDatabasePlugin::close() {
...
+  LOG(INFO) << "repairDB: Deleting corrupted db files";
+  std::vector<std::string> file_names;
+  auto s = listFilesInDirectory(path_, file_names);
+  if (s.ok()) {
+    for (auto file : file_names) {
+      fs::path file_path(file);
+      osquery::remove(file_path);
+    }
+  } else {
+    LOG(INFO) << "Unable to list " << path_ << ". Error " << s.toString();
```
Comment (theopolis) [inline osquery/database/plugins/rocksdb.cpp]: Can you combine this info message into a single phase, without punctuation? For example, "Unable to list /path/to/db: STATUS_STRING".

File: osquery/database/plugins/rocksdb.cpp [hunk 4]
```diff
@@ -201,7 +208,16 @@ Status RocksDBDatabasePlugin::setUp() {
...
       rocksdb::DB::Open(options_, path_, column_families_, &handles_, &db_);
+
+  if (s.IsCorruption()) {
+    // The database is corrupt - try to repair it
+    repairDB();
+    s = rocksdb::DB::Open(options_, path_, column_families_, &handles_, &db_);
+  }
+
   if (!s.ok() || db_ == nullptr) {
+    LOG(INFO) << "Rocksdb open failed [" << s.code() << ":" << s.subcode()
```
Comment (theopolis) [inline osquery/database/plugins/rocksdb.cpp]: Perhaps:
```cpp
LOG(ERROR) << "Cannot open database: " << path_ << " (" << s.code() << ":" << s.subcode() << ") " << s.ToString();
```

I see we use "[" in some places, but () are more common for error codes.

File: osquery/database/plugins/rocksdb.cpp [hunk 5]
```diff
@@ -243,6 +259,22 @@ void RocksDBDatabasePlugin::close() {
   }
 }
 
+void RocksDBDatabasePlugin::repairDB() {
+  // ROCKSDB_LITE does not have a RepairDB method. No option but to delete the
+  // corrupted DB
+  LOG(INFO) << "repairDB: Deleting corrupted db files";
```
Comment (theopolis) [inline osquery/database/plugins/rocksdb.cpp]: nit, "Deleting corrupted database files"

Comment (theopolis) [inline osquery/database/plugins/rocksdb.cpp]: The "repairDB" prefix is not needed.

File: osquery/database/plugins/rocksdb.cpp [hunk 6]
```diff
@@ -243,6 +259,22 @@ void RocksDBDatabasePlugin::close() {
...
+  std::vector<std::string> file_names;
+  auto s = listFilesInDirectory(path_, file_names);
+  if (s.ok()) {
+    for (auto file : file_names) {
+      fs::path file_path(file);
+      osquery::remove(file_path);
```
Comment (theopolis) [inline osquery/database/plugins/rocksdb.cpp]: If you pass a `std::string` it should implicitly convert, right?

**Comment by theopolis** [comment]
ok to test

File: osquery/database/plugins/rocksdb.cpp (hunk 4 above)
Comment (lambda-conjecture) [inline osquery/database/plugins/rocksdb.cpp]: Done

File: osquery/database/plugins/rocksdb.cpp (hunk 5 above)
Comment (lambda-conjecture) [inline osquery/database/plugins/rocksdb.cpp]: Done

File: osquery/database/plugins/rocksdb.cpp (hunk 6 above)
Comment (lambda-conjecture) [inline osquery/database/plugins/rocksdb.cpp]: Done.

File: osquery/database/plugins/rocksdb.cpp (hunk 3 above)
Comment (lambda-conjecture) [inline osquery/database/plugins/rocksdb.cpp]: Done

**Comment by theopolis** [comment]
@lambda-conjecture, is it possible to provide a sample of a corrupted RocksDB database? The `facebook/rocksdb` development team may be able to find out what's causing the corruption.

**Comment by lambda-conjecture** [comment]
Looking into it...

**Comment by lambda-conjecture** [comment]
Quick update - I have reached out to the community where we deployed osquery for help collecting a couple of samples. I had rolled out the fix in this PR to them, so a large number have "recovered", but there are still a bunch that haven't picked up my fix and should have a corrupted DB to look at. I am hopeful we'll get a couple...

**Comment by lambda-conjecture** [comment]
So, unfortunately, at this point, I am not very hopeful of getting a sample :-(. The request is still out to the IT team, so we might yet get a copy. The challenge, of course, is that these are people's work laptops, and for reasons I can only chalk up to a flawed evolutionary pathway,  I am finding that the human population is generally averse to good looking and polite strangers walking up and asking if they can make a copy of a database from their laptops. Somehow, they are much more comfortable with the Orwellian alternative of a script created by an anonymous IT person, runs without their knowledge in the middle of the night and does the exact same thing.

Just wanted to post an update in case it is relevant to making a decision.

**Comment by theopolis** [comment]
The RocksDB developers were very confused, improper shutdown of the database or aggressively killing the daemon should not cause corruption.

That makes me think this solution is more heavy-handed than it already seems. I've implemented a few more 'improvements' to the database handling: (1) periodic refreshing by closing the handling and re-opening, and (2) proper shutdown for worker processes violating their performance limits. We also fixed a potential dead lock with decorators.

My feeling is to let those changes bake for a while. Thoughts?

**Comment by lambda-conjecture** [comment]
From the bit of field experience, I think it would be prudent to take this fix. 

The fix code is only invoked when rocksdb itself returns a status of "isCorrupted", so the proposed fix is pretty conservative. In this situation, the osquery agent becomes completely stranded, so it is a pretty extreme condition. So, I tried to make this as surgical a fix for an admittedly rare but extreme condition. I think the insurance is worth it.

I also wonder if there is interference from other unrelated pieces of software. The same version of osquery has been absolutely flawlessly solid on our linux servers, but had a reasonably consistent occurrence of this issue on end users' mac laptops. Those laptops were all minimally running third party backup software and an anti-virus solution. It sounds plausible that these types of software could interfere with the correct operation of a file-system in some corner-case situation.

Finally, FWIW, there are currently 24 issues against the rocksdb repo related to corruption. At the end of the day, these are complex pieces of software sometimes interconnected in unexpected ways, and operating in a complex and unpredictable environment - which leads me to operate from the point of view that we can all merely conjecture, but never prove, the correctness of the lambda calculus we create  :-)

**Comment by theopolis** [comment]
Totally 👍

**Pull request merged**